metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala

```diff
@@ -189,202 +148,237 @@ class ScalametaLanguageServer(
-    HoverProvider.hover(
-      symbolIndex,
-      Uri(request.params.textDocument),
-      request.params.position.line,
-      request.params.position.character
-    )
-  }
-
-  override def references(
-      request: TextDocumentReferencesRequest
-  ): Task[ReferencesResult] = Task {
-    ReferencesProvider.references(
-      symbolIndex,
-      Uri(request.params.textDocument.uri),
-      request.params.position,
-      request.params.context
-    )
-  }
-
-  override def rename(request: TextDocumentRenameRequest): Task[RenameResult] =
-    Task {
-      RenameProvider.rename(request, symbolIndex, connection)
+    .notification[DidChangeTextDocumentParams](
+      "textDocument/didChange"
+    ) { params =>
+      val changes = params.contentChanges
+      require(changes.length == 1, s"Expected one change, got $changes")
+      val input = Input.VirtualFile(params.textDocument.uri, changes.head.text)
+      buffers.changed(input)
+      sourceChangeSubscriber.onNext(input)
+      ()
     }
-
-  override def signatureHelp(
-      request: TextDocumentSignatureHelpRequest
-  ): Task[SignatureHelpResult] = Task {
-    scalacProvider.getCompiler(request.params.textDocument) match {
-      case Some(g) =>
-        SignatureHelpProvider.signatureHelp(
-          g,
-          toPoint(request.params.textDocument, request.params.position)
-        )
-      case None => SignatureHelpProvider.empty
+    .notification[DidSaveTextDocumentParams](
+      "textDocument/didSave"
+    ) { params =>
+      ()
     }
-  }
-
-  override def onOpenTextDocument(td: TextDocumentItem): Unit = {
-    val input = Input.VirtualFile(td.uri, td.text)
-    buffers.changed(input)
-    sourceChangeSubscriber.onNext(input)
-  }
-
-  override def executeCommand(
-      request: WorkspaceExecuteCommandRequest
-  ): Task[Unit] = Task {
-    import WorkspaceCommand._
-    WorkspaceCommand
-      .withNameOption(request.params.command)
-      .fold(logger.error(s"Unknown command ${request.params.command}")) {
-        case ClearIndexCache =>
+    .notification[DidChangeConfigurationParams](
+      "workspace/didChangeConfiguration"
+    ) { params =>
+      params.settings.hcursor.downField("scalameta").as[Configuration]// match {
+      //   case Left(err) =>
+      //     client.showMessage(MessageType.Error, err.show)
+      //   case Right(conf) =>
+      //     logger.info(s"Configuration updated $conf")
+      //     configurationSubscriber.onNext(conf)
+      // }
+    }
+    .notification[DidChangeWatchedFilesParams](
+      "workspace/didChangeWatchedFiles"
+    ) { params =>
+      params.changes.foreach {
+        case FileEvent(
+            Uri(path),
+            FileChangeType.Created | FileChangeType.Changed
+            ) =>
+          onChangedFile(path.toAbsolutePath) { _ =>
+            logger.warn(s"Unknown file extension for path $path")
+          }
+
+        case event =>
+          logger.warn(s"Unhandled file event: $event")
+          ()
```

Comment:
Doesn’t logger.warn already return unit?

Reply:
I think so, but you better add this comment to @olafurpg's PR. This is just based on it.